accounts/pkg/storage/cs3.go

```diff
@@ -0,0 +1,293 @@
+			},
+		},
+	}, nil
+}
+
+// WriteAccount writes an account via cs3 and modifies the provided account (e.g. with a generated id).
+func (r CS3Repo) WriteAccount(ctx context.Context, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+	if err := r.makeRootDirIfNotExist(ctx, accountsFolder); err != nil {
+		return err
+	}
+
+	var by []byte
+	if by, err = json.Marshal(a); err != nil {
+		return err
+	}
+
+	_, err = r.dataProvider.put(r.accountURL(a.Id), bytes.NewReader(by), t)
+	return err
+}
+
+// LoadAccount loads an account via cs3 by id and writes it to the provided account
+func (r CS3Repo) LoadAccount(ctx context.Context, id string, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.dataProvider.get(r.accountURL(id), t)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return &notFoundErr{"account", id}
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if err := json.Unmarshal(b, &a); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteAccount deletes an account via cs3 by id
+func (r CS3Repo) DeleteAccount(ctx context.Context, id string) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+
+	resp, err := r.storageProvider.Delete(ctx, &provider.DeleteRequest{
+		Ref: &provider.Reference{
+			Spec: &provider.Reference_Path{Path: fmt.Sprintf("/meta/%s/%s", accountsFolder, id)},
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// TODO Handle other error codes?
+	if resp.Status.Code == v1beta11.Code_CODE_NOT_FOUND {
+		return &notFoundErr{"account", id}
+	}
+
+	return nil
+}
+
+// WriteGroup writes a group via cs3 and modifies the provided group (e.g. with a generated id).
+func (r CS3Repo) WriteGroup(ctx context.Context, g *proto.Group) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+	if err := r.makeRootDirIfNotExist(ctx, groupsFolder); err != nil {
+		return err
+	}
+
+	var by []byte
+	if by, err = json.Marshal(g); err != nil {
+		return err
+	}
+
+	_, err = r.dataProvider.put(r.groupURL(g.Id), bytes.NewReader(by), t)
+	return err
+}
+
+// LoadGroup loads a group via cs3 by id and writes it to the provided group
+func (r CS3Repo) LoadGroup(ctx context.Context, id string, g *proto.Group) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.dataProvider.get(r.groupURL(id), t)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return &notFoundErr{"group", id}
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
```

Comment:
The closer interface yields an error, and by deferring the call we're ignoring it. I think it's time to change this for the good and start handling this error. This can be done like:


```
if err := resp. Body.Close(); err != nil {
    log.Fatal(err)
}
```

There is another pattern with a lambda deferred and named return values but it's `magical`. It'd look like this:

```golang
defer func(r http.Response) {
		cerr := r.Body.Close()
		if err == nil {
			err = cerr
		}
	}(*resp)
```

I prefer the first one, simpler.

Reply:
The important bit here is that the body must be closed after being consumed, or a `response body closed` error will pop up.